### ISSUE TITLE
Destructuring of params in arrows uses unbound temporary

### ISSUE THREAD

**Issue by DmitrySoshnikov** [facebook/jstransform#79]
```
errors.sort(({message:[{path:a}]}, {message:[{path:b}]}) => a == b ? 0 : a > b ? -1 : +1);
```

Generated:

```
errors.sort(function($__0, $__1)  {
  var
    $__1 = $__0.message,
    $__2 = $__1[0],
    a = $__2.path,
    $__4 = $__3.message,
    $__5 = $__4[0],
    b = $__5.path;
    return a == b ? 0 : a > b ? -1 : +1;
});
```

There's no `$__3`.

**Comment by zpao** [comment]
It's also reassigning `$__1`. OR the 2nd arg is just  supposed to be `$__3`

**Comment by DmitrySoshnikov** [comment]
Yes, the second param should be `$__3`. The transform in case of params is pretty naive in this case (written by me BTW initially :P, although at that time we didn't have normal temp vars allocator), and just counts the number of structured params to determine the index.

The thing is, params and the body are traversed independently, and it's not known at params processing time how deep a pattern will go inside (i.e. how many temp vars it will need).

**Comment by facebook-github-bot-2** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.
